### PR TITLE
feat(physics): implement repulsion and avoidance forces, make falling of agent possible

### PR DESCRIPTION
**Pull request by kelvin-olaiya**

### CODE REVIEW

File: alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/environments/EnvironmentWithDynamics.kt [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import it.unibo.alchemist.model.interfaces.properties.PhysicalProperty
+import it.unibo.alchemist.model.interfaces.Node.Companion.asProperty
+import org.dyn4j.dynamics.Body
+import org.dyn4j.dynamics.PhysicsBody
+import org.dyn4j.geometry.Circle
+import org.dyn4j.geometry.MassType
+import org.dyn4j.geometry.Transform
+import org.dyn4j.geometry.Vector2
+import org.dyn4j.world.World
+import java.awt.Color
```
Comment (DanySK) [inline alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/environments/EnvironmentWithDynamics.kt:29]: Color???

File: alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/environments/EnvironmentWithDynamics.kt [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+typealias PhysicalProperty2D<T> =
+    PhysicalProperty<T, Euclidean2DPosition, Euclidean2DTransformation, Euclidean2DShapeFactory>
+
+/**
+ * An [Environment] with collision detection and response.
```
Comment (DanySK) [inline alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/environments/EnvironmentWithDynamics.kt:35]: More detail, please :)

File: alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/environments/EnvironmentWithDynamics.kt (hunk 1 above)
Comment (DanySK) [reply alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/environments/EnvironmentWithDynamics.kt:29]: Oh, I see, it loads images.

File: alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/PhysicalPedestrian2D.kt [hunk 3]
```diff
@@ -63,33 +93,89 @@ class PhysicalPedestrian2D<T>(
...
+
+    override fun avoid(other: Node<T>): Euclidean2DPosition {
+        if (environment is Dynamics2DEnvironment) {
+            val distanceVector = (node.position - other.position).let { Vector3D(it.x, it.y, 0.0) }
+            val velocity = environment.getVelocity(node).let { Vector3D(it.x, it.y, 0.0) }
+            val tangentialForce = Vector3D
+                .crossProduct(distanceVector, velocity)
+                .crossProduct(distanceVector)
+                .let {
+                    assert(it.z == 0.0) { "The cross product result should live on the 2D-plane" }
```
Comment (DanySK) [inline alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/PhysicalPedestrian2D.kt:110]:
```suggestion
                    check(it.z == 0.0) { "The cross product result should live on the 2D-plane" }
```

File: alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/PhysicalPedestrian2D.kt [hunk 4]
```diff
@@ -63,33 +93,89 @@ class PhysicalPedestrian2D<T>(
...
+         */
+        private val rectangleOfInfluenceDimensions: Pair<Double, Double> = Pair(3.0, 1.0)
+        /**
+         * Direction tangential force weight factor for when two nodes are moving in the same direction.
+         */
+        private const val directionWeight = 1.2
+        /**
+         * Fallen agent perception radius.
+         */
+        private const val fallenAgentPerceptionRadius = 1.5
```
Comment (DanySK) [inline alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/PhysicalPedestrian2D.kt:179]: Where are these parameters from? And what is their unit?

File: alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/reactions/Physical.kt [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+import it.unibo.alchemist.model.implementations.positions.Euclidean2DPosition
+import it.unibo.alchemist.model.interfaces.Environment
+import it.unibo.alchemist.model.interfaces.Node
+import it.unibo.alchemist.model.interfaces.SteeringAction
+import it.unibo.alchemist.model.interfaces.Time
+import it.unibo.alchemist.model.interfaces.TimeDistribution
+import it.unibo.alchemist.model.interfaces.environments.Dynamics2DEnvironment
+
+/**
+ * A reaction for the update of a [Dynamics2DEnvironment].
```
Comment (DanySK) [inline alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/reactions/Physical.kt:23]: This needs **way** more explanation :smiley: 
What is the idea?

File: alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/reactions/Physical.kt [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+     * Update the internal status of the reaction.
+     */
+    override fun updateInternalStatus(
+        currentTime: Time?,
+        hasBeenExecuted: Boolean,
+        environment: Environment<T, *>?
+    ) = Unit
+
+    /**
+     * Clones this reaction for [node].
```
Comment (DanySK) [inline alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/reactions/Physical.kt:40]:
```suggestion
     * Clones this reaction into [node].
```

File: alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/reactions/Physical.kt (hunk 5 above)
Comment (DanySK) [reply alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/reactions/Physical.kt:23]: In particular, why a reaction and not an action?

File: alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/interfaces/environments/Dynamics2DEnvironment.kt [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+ * as described in the file LICENSE in the Alchemist distribution's top directory.
+ */
+
+package it.unibo.alchemist.model.interfaces.environments
+
+import it.unibo.alchemist.model.implementations.positions.Euclidean2DPosition
+import it.unibo.alchemist.model.interfaces.Node
+
+/**
+ * An [Environment] for physical dynamics.
```
Comment (DanySK) [inline alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/interfaces/environments/Dynamics2DEnvironment.kt:16]: More detail, please

File: alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/interfaces/properties/PhysicalPedestrian.kt [hunk 8]
```diff
@@ -27,6 +27,25 @@ interface PhysicalPedestrian<T, P, A, F> : PhysicalProperty<T, P, A, F>
           A : GeometricTransformation<P>,
           F : GeometricShapeFactory<P, A> {
 
+    /**
+     * Whether the pedestrian is fallen and thus an obstacle.
```
Comment (DanySK) [inline alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/interfaces/properties/PhysicalPedestrian.kt:31]:
```suggestion
     * Whether the pedestrian has fallen and it is thus an obstacle.
```

File: alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/interfaces/properties/PhysicalPedestrian.kt [hunk 9]
```diff
@@ -27,6 +27,25 @@ interface PhysicalPedestrian<T, P, A, F> : PhysicalProperty<T, P, A, F>
...
+     */
+    val isFallen: Boolean
+
+    /**
+     * Turn this [node] into a fallen pedestrian if it [shouldFall].
+     */
+    fun checkAndPossiblyFall()
+
+    /**
+     * Determine if this pedestrian subject to [pushingForces] should fall.
```
Comment (DanySK) [inline alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/interfaces/properties/PhysicalPedestrian.kt:41]:
```suggestion
     * Determines if this pedestrian subject to [pushingForces] should fall.
```

File: alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/interfaces/properties/PhysicalPedestrian.kt [hunk 10]
```diff
@@ -27,6 +27,25 @@ interface PhysicalPedestrian<T, P, A, F> : PhysicalProperty<T, P, A, F>
...
+     * According to the work of [Pelechano et al](https://bit.ly/3e3C7Tb)
+     * this should append when the majority of pushing forces affecting one individual are
```
Comment (DanySK) [inline alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/interfaces/properties/PhysicalPedestrian.kt:43]:
```suggestion
     * this should happen when the majority of pushing forces affecting one individual are
```

File: alchemist-physics/src/test/kotlin/it/unibo/alchemist/test/TestPhysicalPedestrians.kt [hunk 11]
```diff
@@ -10,7 +10,7 @@ import it.unibo.alchemist.testsupport.startSimulation
 class TestPhysicalPedestrians<T, P> : StringSpec({
     "node pushes away a node in the seeking target" {
         loadYamlSimulation<T, P>("pushing_behavior.yml").startSimulation(
-            steps = 300,
+            steps = 800,
```
Comment (DanySK) [inline alchemist-physics/src/test/kotlin/it/unibo/alchemist/test/TestPhysicalPedestrians.kt:13]: why?

File: alchemist-physics/src/test/resources/pushing_behavior.yml [hunk 12]
```diff
@@ -27,7 +26,7 @@ deployments:
       - time-distribution:
           type: DiracComb
           parameters: [1.0]
-        type: BlendedSteeringWithPhysics
+        type: PhysicalBlendedSteering
         actions:
-          - type: CognitiveAgentSeek
+          - type: CognitiveAgentSeek2D
```
Comment (DanySK) [inline alchemist-physics/src/test/resources/pushing_behavior.yml:31]: Do we also have a 3D version?

**Review by DanySK — CHANGES_REQUESTED** [review]
This is nice work.

File: alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/PhysicalPedestrian2D.kt (hunk 4 above)
Comment (kelvin-olaiya) [reply alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/PhysicalPedestrian2D.kt:179]: I got them from [Pelechano et al](https://repository.upenn.edu/cgi/viewcontent.cgi?article=1223&context=hms) paper. In the paper, they're considered in meters.

File: alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/reactions/Physical.kt (hunk 5 above)
Comment (kelvin-olaiya) [reply alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/reactions/Physical.kt:23]: To update the internal `world` I need to know how much time has elapsed. I figured out I can obtain that information from the reaction level. 
Also, I began writing it following what was done for the now removed `BlendedSteeringWithPhysics`. I needed the `steeringStrategy` computation to obtain the next step position, taking into consideration the physical forces.

File: alchemist-physics/src/test/resources/pushing_behavior.yml (hunk 12 above)
Comment (kelvin-olaiya) [reply alchemist-physics/src/test/resources/pushing_behavior.yml:31]: We do not have a 3D version. I switched to the 2D as it was documented having more sophisticated  movements.

File: alchemist-physics/src/test/kotlin/it/unibo/alchemist/test/TestPhysicalPedestrians.kt (hunk 11 above)
Comment (kelvin-olaiya) [reply alchemist-physics/src/test/kotlin/it/unibo/alchemist/test/TestPhysicalPedestrians.kt:13]: When the test was written, there were no forces other than the repulsion force. I increased the number of step as it took a little longer to push the node away from its position.

Comment (kelvin-olaiya) [reply alchemist-physics/src/test/kotlin/it/unibo/alchemist/test/TestPhysicalPedestrians.kt:13]: I'm just noticing the test is no longer passing anyway. I'll try to figure out why...

File: alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/PhysicalPedestrian2D.kt (hunk 4 above)
Comment (DanySK) [reply alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/PhysicalPedestrian2D.kt:179]: Then write both the source and the unit in the kdoc, please :)

File: alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/reactions/Physical.kt (hunk 5 above)
Comment (DanySK) [reply alchemist-physics/src/main/kotlin/it/unibo/alchemist/model/implementations/reactions/Physical.kt:23]: ok

**Review by DanySK — APPROVED** [review]

**Pull request merged**